federation/pkg/federation-controller/dns/dns.go

```diff
@@ -175,179 +303,109 @@ func getResolvedEndpoints(endpoints []string) ([]string, error) {
 	return resolvedEndpoints, nil
 }
 
-/* ensureDnsRrsets ensures (idempotently, and with minimum mutations) that all of the DNS resource record sets for dnsName are consistent with endpoints.
-   if endpoints is nil or empty, a CNAME record to uplevelCname is ensured.
-*/
-func (s *ServiceController) ensureDnsRrsets(dnsZone dnsprovider.Zone, dnsName string, endpoints []string, uplevelCname string) error {
+// ensureDNSRrsets ensures (idempotently, and with minimum mutations) that all of the DNS resource record sets for
+// dnsName are consistent with endpoints. if endpoints is nil or empty, a CNAME record to uplevelCname is ensured.
+func (dc *DNSController) ensureDNSRrsets(dnsZone dnsprovider.Zone, service *apiv1.Service, addRecords bool) (
```

Comment:
I was trying to ensure all the dns entries are updated once per service instead of multiple times (service * no.of_clusters). it will reduce total number of operations per service, if multiple global endpoints needs to be updated and also the changeset.Apply() will be called only once per service.
I would update the code and try to remove the coupling with service for this function.